File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/review/ReviewEditViewModel.kt
package com.hhp227.concafe.presentation.review

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.concafe.domain.common.AppError
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.usecase.CreateReviewUseCase
import com.hhp227.concafe.domain.usecase.GetCafeDetailUseCase
import com.hhp227.concafe.domain.usecase.GetReviewUseCase
import com.hhp227.concafe.domain.usecase.UpdateReviewUseCase
import com.hhp227.concafe.domain.usecase.UploadImageUseCase
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class ReviewEditViewModel(
    private val cafeId: String? = null,
    private val reviewId: String? = null,
    private val getCafeDetailUseCase: GetCafeDetailUseCase,
    private val createReviewUseCase: CreateReviewUseCase,
    private val updateReviewUseCase: UpdateReviewUseCase,
    private val getReviewUseCase: GetReviewUseCase,
    private val uploadImageUseCase: UploadImageUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow(
        ReviewEditUiState(
            cafeId = cafeId.orEmpty(),
            reviewId = reviewId,
            screenTitle = if (reviewId != null) "리뷰 수정" else "리뷰 작성",
            topActionLabel = if (reviewId != null) "수정" else "등록",
            submitButtonLabel = if (reviewId != null) "리뷰 수정하기" else "리뷰 등록하기"
        )
    )
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<ReviewEditEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private fun loadCafeInfo() {
        if (cafeId.isNullOrBlank()) {
            _uiState.update { it.copy(infoMessage = "카페 정보를 찾을 수 없습니다.") }
        } else {
            _uiState.update {
                it.copy(
                    isLoading = true,
                    infoMessage = null
                )
            }
            viewModelScope.launch {
                when (val result = getCafeDetailUseCase.invoke(cafeId)) {
                    is AppResult.Success -> {
                        val detail = result.data.detail
                        _uiState.update {
                            it.copy(
                                isLoading = false,
                                cafeId = cafeId,
                                cafeName = detail.cafe.name,
                                cafeAddress = detail.cafe.region.address,
                                isLoggedIn = result.data.isLoggedIn,
                                isVisitVerified = result.data.isVisitVerified,
                                availableCastTags = detail.casts.map { cast ->
                                    ReviewEditUiState.CastTag(
                                        id = cast.id,
                                        name = cast.name
                                    )
                                },
                                infoMessage = null
                            )
                        }
                        if (reviewId != null) loadExistingReview()
                    }
                    is AppResult.Failure -> {
                        _uiState.update {
                            it.copy(
                                isLoading = false,
                                infoMessage = "카페 정보를 불러오지 못했습니다."
                            )
                        }
                    }
                }
            }
        }
    }

    private fun loadExistingReview() {
        val id = reviewId ?: return
        viewModelScope.launch {
            when (val result = getReviewUseCase.invoke(id)) {
                is AppResult.Success -> {
                    val review = result.data
                    _uiState.update {
                        it.copy(
                            rating = review.rating.toInt(),
                            content = review.content,
                            taggedCastIds = review.taggedCastIds,
                            photoImageUrl = review.imageUrls.firstOrNull()
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update { it.copy(infoMessage = "기존 리뷰를 불러오지 못했습니다.") }
                }
            }
        }
    }

    private fun clickBack() {
        viewModelScope.launch {
            _event.emit(ReviewEditEvent.NavigateBack)
        }
    }

    private fun clickAddPhoto() {
        _uiState.update { it.copy(infoMessage = null) }
    }

    private fun removePhoto() {
        _uiState.update { it.copy(photoImageUrl = null) }
    }

    private fun toggleCastTag(castId: String) {
        _uiState.update { state ->
            val nextIds = state.taggedCastIds.toMutableList().apply {
                if (contains(castId)) {
                    remove(castId)
                } else {
                    add(castId)
                }
            }
            state.copy(taggedCastIds = nextIds)
        }
    }

    private fun clickSubmit() {
        val currentState = _uiState.value

        if (currentState.rating <= 0) {
            _uiState.update { it.copy(infoMessage = "평점을 선택해주세요.") }
        } else if (currentState.content.trim().length < ReviewEditUiState.minimumReviewLength) {
            _uiState.update { it.copy(infoMessage = "상세 리뷰는 최소 10자 이상 입력해주세요.") }
        } else {
            _uiState.update {
                it.copy(
                    isSubmitting = true,
                    infoMessage = null
                )
            }
            viewModelScope.launch {
                val uploadedPhoto = uploadImage(currentState.photoImageUrl, "reviews")
                if (!currentState.photoImageUrl.isNullOrBlank() && uploadedPhoto == null) {
                    return@launch
                }
                val imageUrls = uploadedPhoto?.let { listOf(it) } ?: emptyList()

                if (reviewId != null) {
                    when (
                        val result = updateReviewUseCase.invoke(
                            reviewId = reviewId,
                            rating = currentState.rating.toFloat(),
                            content = currentState.content,
                            imageUrls = imageUrls,
                            taggedCastIds = currentState.taggedCastIds
                        )
                    ) {
                        is AppResult.Success -> {
                            _uiState.update { it.copy(isSubmitting = false) }
                            _event.emit(ReviewEditEvent.NavigateBack)
                        }
                        is AppResult.Failure -> {
                            _uiState.update {
                                it.copy(
                                    isSubmitting = false,
                                    infoMessage = when (val error = result.error) {
                                        is AppError.Unauthorized -> "리뷰 수정은 로그인 후 가능해요."
                                        is AppError.ValidationFailed -> error.reason.toReviewValidationMessage()
                                        else -> "리뷰 수정에 실패했습니다."
                                    }
                                )
                            }
                        }
                    }
                } else {
                    when (
                        val result = createReviewUseCase.invoke(
                            cafeId = currentState.cafeId,
                            rating = currentState.rating.toFloat(),
                            content = currentState.content,
                            imageUrls = imageUrls,
                            taggedCastIds = currentState.taggedCastIds
                        )
                    ) {
                        is AppResult.Success -> {
                            _uiState.update {
                                it.copy(
                                    isSubmitting = false,
                                    reviewId = result.data.id,
                                    userId = result.data.userId,
                                    visitId = result.data.visitId,
                                    createdAt = result.data.createdAt
                                )
                            }
                            _event.emit(ReviewEditEvent.NavigateBack)
                        }
                        is AppResult.Failure -> {
                            _uiState.update {
                                it.copy(
                                    isSubmitting = false,
                                    infoMessage = when (val error = result.error) {
                                        is AppError.Unauthorized -> "리뷰 작성은 로그인 후 가능해요."
                                        is AppError.ValidationFailed -> error.reason.toReviewValidationMessage()
                                        else -> "리뷰 등록에 실패했습니다."
                                    }
                                )
                            }
                        }
                    }
                }
            }
        }
    }

    fun onAction(action: ReviewEditAction) {
        when (action) {
            ReviewEditAction.ClickBack -> clickBack()
            is ReviewEditAction.SelectRating -> _uiState.update {
                it.copy(
                    rating = action.rating.coerceIn(0, ReviewEditUiState.maximumRating)
                )
            }
            ReviewEditAction.ClickAddPhoto -> clickAddPhoto()
            is ReviewEditAction.SelectPhoto -> _uiState.update { it.copy(photoImageUrl = action.imageUrl, infoMessage = null) }
            ReviewEditAction.RemovePhoto -> removePhoto()
            is ReviewEditAction.ChangeReviewText -> _uiState.update {
                it.copy(content = action.value)
            }
            is ReviewEditAction.ToggleCastTag -> toggleCastTag(action.castId)
            is ReviewEditAction.SelectAtmosphereAnswer -> _uiState.update {
                it.copy(atmosphereAnswer = action.isPositive)
            }
            ReviewEditAction.ClickSubmit -> clickSubmit()
            ReviewEditAction.DismissInfoMessage -> _uiState.update { it.copy(infoMessage = null) }
        }
    }

    private suspend fun uploadImage(imageUrl: String?, folder: String): String? {
        if (imageUrl.isNullOrBlank()) return null
        return when (val result = uploadImageUseCase.invoke(imageUrl, folder)) {
            is AppResult.Success -> result.data
            is AppResult.Failure -> {
                _uiState.update {
                    it.copy(
                        isSubmitting = false,
                        infoMessage = "이미지를 업로드하지 못했습니다."
                    )
                }
                null
            }
        }
    }

    init {
        loadCafeInfo()
    }
}

private fun String.toReviewValidationMessage(): String {
    return when (this) {
        "cafeId is required" -> "카페 정보를 찾을 수 없습니다."
        "rating is required" -> "평점을 선택해주세요."
        "review content is required" -> "상세 리뷰를 입력해주세요."
        else -> this
    }
}

File: iosApp/iosApp/Presentation/Review/ReviewEditViewModel.swift
//
//  ReviewEditViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/12.
//

import Foundation
import Combine
import Shared

@MainActor
final class ReviewEditViewModel: ObservableObject {
    @Published private(set) var uiState = ReviewEditUiState()

    let event = PassthroughSubject<ReviewEditEvent, Never>()

    private let cafeId: String?

    private let getCafeDetailUseCase: GetCafeDetailUseCase

    private let createReviewUseCase: CreateReviewUseCase

    private let uploadImageUseCase: UploadImageUseCase

    private var loadTask: Task<Void, Never>?

    private func loadCafeInfo() {
        if let cafeId, !cafeId.isEmpty {
            loadTask?.cancel()
            uiState.isLoading = true
            uiState.infoMessage = nil
            uiState.cafeId = cafeId

            loadTask = Task {
                do {
                    let result = try await getCafeDetailUseCase.invoke(cafeId: cafeId)

                    if let success = result as? AppResultSuccess<AnyObject>,
                       let feed = success.data as? CafeDetailFeed {
                        let detail = feed.detail
                        uiState.isLoading = false
                        uiState.cafeId = cafeId
                        uiState.cafeName = detail.cafe.name
                        uiState.cafeAddress = detail.cafe.region.address
                        uiState.isLoggedIn = feed.isLoggedIn
                        uiState.isVisitVerified = feed.isVisitVerified
                        uiState.availableCastTags = detail.casts.map { cast in
                            ReviewEditUiState.CastTag(id: cast.id, name: cast.name)
                        }
                        uiState.infoMessage = nil
                    } else {
                        uiState.isLoading = false
                        uiState.infoMessage = "카페 정보를 불러오지 못했습니다."
                    }
                } catch {
                    if Task.isCancelled { return }
                    uiState.isLoading = false
                    uiState.infoMessage = "카페 정보를 불러오지 못했습니다."
                }
            }
        } else {
            uiState.infoMessage = "카페 정보를 찾을 수 없습니다."
        }
    }

    private func clickAddPhoto() {
        uiState.infoMessage = nil
    }

    private func removePhoto() {
        uiState.photoImageUrl = nil
    }

    private func toggleCastTag(_ castId: String) {
        if uiState.taggedCastIds.contains(castId) {
            uiState.taggedCastIds.removeAll { currentId in
                currentId == castId
            }
        } else {
            uiState.taggedCastIds.append(castId)
        }
    }

    private func clickSubmit() {
        let reviewLength = uiState.content.trimmingCharacters(in: .whitespacesAndNewlines).count

        if uiState.rating <= 0 {
            uiState.infoMessage = "평점을 선택해주세요."
        } else if reviewLength < ReviewEditUiState.minimumReviewLength {
            uiState.infoMessage = "상세 리뷰는 최소 10자 이상 입력해주세요."
        } else {
            uiState.isSubmitting = true
            uiState.infoMessage = nil

            loadTask?.cancel()
            loadTask = Task {
                do {
                    let uploadedPhoto = try await uploadImageIfNeeded(uiState.photoImageUrl, folder: "reviews")
                    let result = try await createReviewUseCase.invoke(
                        cafeId: uiState.cafeId,
                        rating: Float(uiState.rating),
                        content: uiState.content,
                        imageUrls: uploadedPhoto.map { [$0] } ?? [],
                        taggedCastIds: uiState.taggedCastIds
                    )

                    if let success = result as? AppResultSuccess<AnyObject> {
                        uiState.isSubmitting = false
                        if let review = success.data as? Review {
                            uiState.reviewId = review.id
                            uiState.userId = review.userId
                            uiState.visitId = review.visitId
                            uiState.createdAt = review.createdAt
                        }
                        event.send(.navigateBack)
                    } else if let failure = result as? AppResultFailure {
                        uiState.isSubmitting = false
                        if failure.error is AppErrorUnauthorized {
                            uiState.infoMessage = "리뷰 작성은 로그인 후 가능해요."
                        } else if let error = failure.error as? AppErrorValidationFailed {
                            uiState.infoMessage = Self.reviewValidationMessage(for: error.reason)
                        } else {
                            uiState.infoMessage = "리뷰 등록에 실패했습니다."
                        }
                    } else {
                        uiState.isSubmitting = false
                        uiState.infoMessage = "리뷰 등록에 실패했습니다."
                    }
                } catch {
                    if Task.isCancelled { return }
                    uiState.isSubmitting = false
                    uiState.infoMessage = "리뷰 등록에 실패했습니다."
                }
            }
        }
    }

    func onAction(_ action: ReviewEditAction) {
        switch action {
        case .clickBack:
            event.send(.navigateBack)
        case .selectRating(let rating):
            uiState.rating = min(max(rating, 0), ReviewEditUiState.maximumRating)
        case .clickAddPhoto:
            clickAddPhoto()
        case .selectPhoto(let imageUrl):
            uiState.photoImageUrl = imageUrl
            uiState.infoMessage = nil
        case .removePhoto:
            removePhoto()
        case .changeReviewText(let value):
            uiState.content = value
        case .toggleCastTag(let castId):
            toggleCastTag(castId)
        case .selectAtmosphereAnswer(let isPositive):
            uiState.atmosphereAnswer = isPositive
        case .clickSubmit:
            clickSubmit()
        case .dismissInfoMessage:
            uiState.infoMessage = nil
        }
    }

    private static func reviewValidationMessage(for reason: String) -> String {
        switch reason {
        case "cafeId is required":
            return "카페 정보를 찾을 수 없습니다."
        case "rating is required":
            return "평점을 선택해주세요."
        case "review content is required":
            return "상세 리뷰를 입력해주세요."
        default:
            return reason
        }
    }

    init(
        cafeId: String? = nil,
        getCafeDetailUseCase: GetCafeDetailUseCase = KoinInitializerKt.resolveGetCafeDetailUseCase(),
        createReviewUseCase: CreateReviewUseCase = KoinInitializerKt.resolveCreateReviewUseCase(),
        uploadImageUseCase: UploadImageUseCase = KoinInitializerKt.resolveUploadImageUseCase()
    ) {
        self.cafeId = cafeId
        self.getCafeDetailUseCase = getCafeDetailUseCase
        self.createReviewUseCase = createReviewUseCase
        self.uploadImageUseCase = uploadImageUseCase
        uiState.cafeId = cafeId ?? ""

        loadCafeInfo()
    }

    deinit {
        loadTask?.cancel()
    }

    private func uploadImageIfNeeded(_ imageUrl: String?, folder: String) async throws -> String? {
        guard let imageUrl, !imageUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }
        let result = try await uploadImageUseCase.invoke(localPath: imageUrl, folder: folder)
        if let success = result as? AppResultSuccess<AnyObject>, let data = success.data as? String {
            return data
        }
        if let failure = result as? AppResultFailure, let validation = failure.error as? AppErrorValidationFailed {
            throw NSError(domain: "ReviewEdit", code: 1, userInfo: [NSLocalizedDescriptionKey: validation.reason])
        }
        throw NSError(domain: "ReviewEdit", code: 1, userInfo: [NSLocalizedDescriptionKey: "이미지를 업로드하지 못했습니다."])
    }
}
